File: CommandPython/CeilingFanCommand.py
# Copyright (c) 2022 RodKingroo

from CeilingFan import *
from Command import *

def setUndo(ceilingFan, prevSpeed):
    if prevSpeed == ceilingFan.OFF:
        ceilingFan.off()
    elif prevSpeed == ceilingFan.STAGE_ONE:
        ceilingFan.stage_1()
    elif prevSpeed == ceilingFan.STAGE_TWO:
        ceilingFan.stage_2()
    elif prevSpeed == ceilingFan.STAGE_THREE:
        ceilingFan.stage_3()
    elif prevSpeed == ceilingFan.STAGE_FOUR:
        ceilingFan.stage_4()
    elif prevSpeed == ceilingFan.STAGE_FIVE:
        ceilingFan.stage_5()
    return ceilingFan

# Ceiling Fan Off Command
class CeilingFanOffCommand(Command):
    def __init__(self, ceilingFan):
        if isinstance(ceilingFan, CeilingFan):
            self.ceilingFan = ceilingFan
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevSpeed = self.ceilingFan.getState()
        self.ceilingFan.off()
        
    def undo(self):
        self.ceilingFan.on()
        self.ceilingFan = setUndo(self.ceilingFan, self.prevSpeed)

# Ceiling Fan Stage One Command
class CeilingFanStageOneCommand(Command):
    def __init__(self, ceilingFan):
        if isinstance(ceilingFan, CeilingFan):
            self.ceilingFan = ceilingFan
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevSpeed = self.ceilingFan.getState()
        self.ceilingFan.on()
        self.ceilingFan.stage_1()
        
    def undo(self):
        self.ceilingFan.off()
        self.ceilingFan = setUndo(self.ceilingFan, self.prevSpeed)
        
        
# Ceiling Fan Stage Two Command
class CeilingFanStageTwoCommand(Command):
    def __init__(self, ceilingFan):
        if isinstance(ceilingFan, CeilingFan):
            self.ceilingFan = ceilingFan
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevSpeed = self.ceilingFan.getState()
        self.ceilingFan.on()
        self.ceilingFan.stage_2()    
        
    def undo(self):
        self.ceilingFan.off()
        self.ceilingFan = setUndo(self.ceilingFan, self.prevSpeed)
   
        
# Ceiling Fan Stage Three Command
class CeilingFanStageThreeCommand(Command):
    def __init__(self, ceilingFan):
        if isinstance(ceilingFan, CeilingFan):
            self.ceilingFan = ceilingFan
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevSpeed = self.ceilingFan.getState()
        self.ceilingFan.on()
        self.ceilingFan.stage_3()
        
    def undo(self):
        self.ceilingFan.off()
        self.ceilingFan = setUndo(self.ceilingFan, self.prevSpeed)
        
        
# Ceiling Fan Stage Four Command
class CeilingFanStageFourCommand(Command):
    def __init__(self, ceilingFan):
        if isinstance(ceilingFan, CeilingFan):
            self.ceilingFan = ceilingFan
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevSpeed = self.ceilingFan.getState()
        self.ceilingFan.on()
        self.ceilingFan.stage_4()
        
    def undo(self):
        self.ceilingFan.off()
        self.ceilingFan = setUndo(self.ceilingFan, self.prevSpeed)
        

# Ceiling Fan Stage Five Command
class CeilingFanStageFiveCommand(Command):
    def __init__(self, ceilingFan):
        if isinstance(ceilingFan, CeilingFan):
            self.ceilingFan = ceilingFan
        else:
            raise NotImplementedError
    
    def execute(self):
        self.prevSpeed = self.ceilingFan.getState()
        self.ceilingFan.on()
        self.ceilingFan.stage_5()
        
    def undo(self):
        self.ceilingFan.off()
        self.ceilingFan = setUndo(self.ceilingFan, self.prevSpeed)

File: CommandCPP/CeilingFanCommand.cpp
#include "CeilingFanCommand.h"

void setUndo(shared_ptr<CeilingFan> ceilingFan, int prevSpeed)
{
    switch (prevSpeed)
    {
    case CeilingFan::OFF:
        ceilingFan->off();
        break;
    case CeilingFan::STATE_ONE:
        ceilingFan->state_1();
        break;
    case CeilingFan::STATE_TWO:
        ceilingFan->state_2();
        break;
    case CeilingFan::STATE_THREE:
        ceilingFan->state_3();
        break;
    case CeilingFan::STATE_FOUR:
        ceilingFan->state_4();
        break;
    case CeilingFan::STATE_FIVE:
        ceilingFan->state_5();
        break;

    }
}


/* Ceiling Fan One State Command */
CeilingFanOneStateCommand::CeilingFanOneStateCommand(shared_ptr<CeilingFan> ceilingFan)
{
    this->ceilingFan = ceilingFan;
}

void CeilingFanOneStateCommand::execute() /* override */
{
    prevSpeed = ceilingFan->getState(); /// CeilingFan.h
    ceilingFan->on(); 
    ceilingFan->state_1();

}

void CeilingFanOneStateCommand::undo() /* override */
{
    ceilingFan->off(); /// CeilingFan.h
    setUndo(ceilingFan, prevSpeed);
}


/* Ceiling Fan Two State Command */
CeilingFanTwoStateCommand::CeilingFanTwoStateCommand(shared_ptr<CeilingFan> ceilingFan)
{
    this->ceilingFan = ceilingFan;
}

void CeilingFanTwoStateCommand::execute() /* override */
{
    prevSpeed = ceilingFan->getState(); // CeilingFan.h
    ceilingFan->on();
    ceilingFan->state_2();

}

void CeilingFanTwoStateCommand::undo() /* override */
{
    ceilingFan->off(); //CeilingFan.h
    setUndo(ceilingFan, prevSpeed);
}


/* Ceiling Fan Three State Command */
CeilingFanThreeStateCommand::CeilingFanThreeStateCommand(shared_ptr<CeilingFan> ceilingFan)
{
    this->ceilingFan = ceilingFan;    
}

void CeilingFanThreeStateCommand::execute() /* override */
{
    prevSpeed = ceilingFan->getState();
    ceilingFan->on();
    ceilingFan->state_3();

}

void CeilingFanThreeStateCommand::undo() /* override */
{
    ceilingFan->off();
    setUndo(ceilingFan, prevSpeed);

}


/* Ceiling Fan Four State Command */
CeilingFanFourStateCommand::CeilingFanFourStateCommand(shared_ptr<CeilingFan> ceilingFan)
{
    this->ceilingFan = ceilingFan;

}

void CeilingFanFourStateCommand::execute() /* override */
{
    prevSpeed = ceilingFan->getState();
    ceilingFan->on();
    ceilingFan->state_4();

}

void CeilingFanFourStateCommand::undo() /* override */
{
    ceilingFan->off();
    setUndo(ceilingFan, prevSpeed);
}


/*Ceiling Fan Five State Command*/
CeilingFanFiveStateCommand::CeilingFanFiveStateCommand(shared_ptr<CeilingFan> ceilingFan)
{
    this->ceilingFan = ceilingFan;

}

void CeilingFanFiveStateCommand::execute() /* override */
{
    prevSpeed = ceilingFan->getState();
    ceilingFan->on();
    ceilingFan->state_5();

}

void CeilingFanFiveStateCommand::undo() /* override */
{
    ceilingFan->off();
    setUndo(ceilingFan, prevSpeed);

}


/* Ceiling Fan Off Command*/
CeilingFanOffCommand::CeilingFanOffCommand(shared_ptr<CeilingFan> ceilingFan)
{
    this->ceilingFan = ceilingFan;
}

void CeilingFanOffCommand::execute() /* override */
{
    prevSpeed = ceilingFan->getState();
    ceilingFan->off();

}

void CeilingFanOffCommand::undo() /* override */
{
    ceilingFan->on();
    setUndo(ceilingFan, prevSpeed);
}
